src/cell_list.h

```diff
@@ -0,0 +1,226 @@
+#pragma once
+#ifndef CELL_LIST_H
+#define CELL_LIST_H
+
+#include <algorithm>
+#include <list>
+#include <vector>
+
+#include "cata_utility.h"
+#include "line.h"
+#include "point.h"
+
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_iterator;
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_range;
+
+/*
+ * A Cell List operates by subdividing an area into a regular grid and populating each cell
+ * with a list of the entities in that area.
+ * It is used to optimize lookup of nearby entities.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list
+{
+    public:
+        cell_list() = default;
+        void add( const point &p, T value ) {
+            std::pair<point, T> val( p, value );
+            cells[p.x / CELL_SIZE][p.y / CELL_SIZE].push_back( val );
+        }
+        cell_list_range< T, CELL_SIZE, MAXX, MAXY > get( const point &p, int distance ) {
```

Comment:
Can it become `const`?

Reply:
Sort of? I could mark sorted_elements and friends mutable, but that's most of the data structure so it doesn't seem worthwhile.